[WabiSabi] GetStatus API and End-to-End integration test as template.

lontivero opened the pull request:
Allows to fetch information about the existing rounds from the coordinator. To do this we created a minimal `RoundState` record containing the coordinators parameters, fee rate and coinjoin state. This class is very similar to the one added by @molnard in https://github.com/zkSNACKs/WalletWasabi/pull/5694.

Given `RoundState` includes the `IState` (`ContructionState` or `SigningState`) a  new `IStateJsonConverter` was needed. I tried really hard to use what Newtonsoft provides but I found lots of problems and conflict with other already existing json converteres). Additionally other converters were needed.

Having the `http://<host>/wabisabi/status` endpoint available allowed us to finally create a good end-to-end (client, serialization,network,deserialization,process,serialization,exception handling, deserialization) test. These tests run incredibly fast and are really flexible.

Finally, I renamed the `IArenaRequestHandler` as `IWabiSabiApiRequestHandler` and made the api controller     to implement it instead of the `ArenaRequestHandler`.

nothingmuch commented on WalletWasabi.Backend/Filters/ExceptionTranslateFilter.cs:
```diff
@@ -0,0 +1,35 @@
+					Exception e => new Error(
+						Type: "unknown",
+						ErrorCode: "0",
+						Description: e.Message)
```
exceptions other than `WabiSabiProtocolException` should not be sent to the clients, but only to the server logs.

nothingmuch commented on WalletWasabi/WabiSabi/Models/RoundState.cs:
```diff
@@ -0,0 +1,18 @@
+		uint256 Id,
+		CredentialIssuerParameters AmountCredentialIssuerParameters,
+		CredentialIssuerParameters VsizeCredentialIssuerParameters,
+		FeeRate FeeRate,
```
Feerate is already in `MultipartyTransactionParameters` (sorry, I still need to finish cleanups relating to that). Note that when encryption is reintroduced, that will need to be serialized separately from the state data, because that would need to remain in cleartext.

nothingmuch commented on WalletWasabi.Tests/RegressionTests/WabiSabi/Startup.cs:
```diff
@@ -0,0 +1,37 @@
+using Microsoft.Extensions.DependencyInjection;
+using WalletWasabi.WabiSabi.Models.Serialization;
+
+namespace WalletWasabi.Tests.RegressionTests.WabiSabi
```
why `RegressionTests`? to me that term means testing to ensure that regressions don't happen, i.e. something that was previously working stops working (in practice, usually added as part of fixing a regression that was missed by tests that were in place at the time that the regression was accidentally introduced)

(sorry, i missed this in the PR that introduced `RegressionTests/WabiSabi`)

nothingmuch reviewed (commented):
a few comments, but i only skimmed this PR, still need to do a more detailed read

lontivero commented on WalletWasabi.Tests/RegressionTests/WabiSabi/WabiSabiApiApplicationFactory.cs:
```diff
@@ -7,92 +7,64 @@
+				services.AddScoped<IRPCClient>(_ => BitcoinFactory.GetMockMinimalRpc());
+				services.AddScoped<Prison>();
+				services.AddScoped<WabiSabiConfig>();
+				services.AddScoped(typeof(TimeSpan), _ => TimeSpan.FromSeconds(2));
```
This `Startup` class is for setting up the testing http server only. It has nothing to do with the  `WalletWasabi.Backend.Setup` class that is the one that runs in real world. However it is really useful to understand what has to be setted and how, because we can see what dependencies are required, how to register hosted services, how to register the json converters and how to inject value types, and who knows what else we can have in the future.

Currently, instead of doing things in this way, that is how we should register all the things, we have a `WabiSabiCoordinator` class that does what we are doing here.

-----------------

Going back to your question: in real world we should read the value from the `WabiSabiConfig` and inject that.

lontivero commented:
> Am I doing something wrong?

No, your are doing it right. The problem is the tests were broken after merging master into this branch and the tests you ran where not running as part of the CI. 

I deleted the old low-quality tests that didn't failed and created a new integration test that verifies the correct behavior and finally i moved the tests to make it run always (they are really fast so, there will be no problems).

yahiheb commented on WalletWasabi.Tests/Helpers/HttpRequestMessageExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+using System.Net.Http;
+
+namespace WalletWasabi.Tests.UnitTests.WabiSabi.Integration
```
namespace should be `WalletWasabi.Tests/Helpers`

yahiheb reviewed (commented):
Only nits.

molnard commented on WalletWasabi.Backend/Filters/IdempotentActionFilterAttributeImpl.cs:
```diff
@@ -18,32 +19,36 @@ public IdempotentActionFilterAttributeImpl(IMemoryCache cache)
 			var request = context.HttpContext.Request;
 
-			var (cacheKey, body) = GetCacheEntryKey(request);
+			var (cacheKey, body) = await GetCacheEntryKeyAsync(request);
```
ConfigureAwait(false)?

lontivero replied:
These are server-side async tasks and in asp.net core at least there is no custom synchronization context, that's why we don't use it anywhere in the backend. Anyway, I think it is okay to use it in the backend also.

molnard commented on WalletWasabi.Backend/Filters/IdempotentActionFilterAttributeImpl.cs:
```diff
@@ -18,32 +19,36 @@ public IdempotentActionFilterAttributeImpl(IMemoryCache cache)
 
 			context.HttpContext.Items["cached-key"] = cacheKey;
+
+ 			await next();
```
ConfigureAwait(false)?

molnard commented on WalletWasabi.Backend/Filters/IdempotentActionFilterAttributeImpl.cs:
```diff
@@ -18,32 +19,36 @@ public IdempotentActionFilterAttributeImpl(IMemoryCache cache)
 		{
 			using var reader = new System.IO.StreamReader(request.Body);
-			var body = reader.ReadToEnd();
+			var body = await reader.ReadToEndAsync();
```
ConfigureAwait(false)?

molnard reviewed (changes requested):
LGTM

molnard commented on WalletWasabi.Backend/Filters/ExceptionTranslateFilter.cs:
```diff
@@ -0,0 +1,31 @@
+		public override void OnException(ExceptionContext context)
+		{
+			var exception = context.Exception.InnerException ?? context.Exception;
+			var exceptionMessage = exception.Message;
```
This is not used.

molnard commented on WalletWasabi/Tor/Http/Extensions/HttpResponseMessageExtensions.cs:
```diff
@@ -61,6 +66,23 @@ public static async Task ThrowRequestExceptionFromContentAsync(this HttpResponse
 			if (me.Content is { })
 			{
 				var contentString = await me.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+				var error = JsonConvert.DeserializeObject<Error>(contentString, new JsonSerializerSettings()
```
What happens if it cannot deserialize? If it throws then we will never get to the line of further processing here: https://github.com/zkSNACKs/WalletWasabi/pull/5767/files#diff-692d4a4f7d2eeacc2012f438873c1b1d2e98cb92db69dd0592df35644d71c96fR88

lontivero replied:
It doesn't throw because the `Error` handler ignores any problem and makes sure to return `null` in case the deserialization is not possible.

molnard commented on WalletWasabi/WabiSabi/Models/Serialization/IStateJsonConverter.cs:
```diff
@@ -0,0 +1,98 @@
+
+namespace WalletWasabi.WabiSabi.Models.Serialization
+{
+	public class IStateJsonConverter : JsonConverter
```
nit: Raname to IRoundStateJsonConverter ?

lontivero replied:
Ummm the problem is not this json converter's name, the problem is the `IState` interface. 

I don't want to call this converter `RoundStateJsonConverter` because that would mean that the converter works with the class called `RoundState`, what it does not. This converter converts `IState` instances (`ConstructionState` and `SigningState`).

It would be good to rename that interface to something better and then rename this converter too but `RoundStateJsonConverter` would be misleading at best.

nothingmuch replied:
:+1: renaming `IState`, consistent with params would be `IMultipartyTransactionState`, maybe a bit of a mouthful, but clear.

bear in mind that when we reintroduce encryption we will have to revisit this, and I think `IState` will not be serialized directly since the parameters have to be unencrypted but the events/state transitions have to be encrypted, and we probably want to serialize the encrypted data using `IBitcoinSerializable` since it's almost entirely pieces of the TX anyway.

molnard reviewed (changes requested):
Final review
